source/beamline/BioXAS/BioXASSSRLMonochromatorRegionControl.cpp

```diff
@@ -0,0 +1,846 @@
+#include "BioXASSSRLMonochromatorRegionControl.h"
+#include "BioXASSSRLMonochromator.h"
+
+BioXASSSRLMonochromatorRegionControl::BioXASSSRLMonochromatorRegionControl(QObject *parent) :
+	AMCompositeControl("RegionControl", "", parent, "BioXAS SSRL Monochromator Region Control")
+{
+	// Initialize local variables.
+
+	value_ = BioXASSSRLMonochromator::Region::None;
+	setpoint_ = BioXASSSRLMonochromator::Region::None;
+	moveInProgress_ = false;
+
+	upperSlit_ = 0;
+	lowerSlit_ = 0;
+	slitsStatus_ = 0;
+	paddle_ = 0;
+	paddleStatus_ = 0;
+	keyStatus_ = 0;
+	brakeStatus_ = 0;
+	bragg_ = 0;
+	braggAtCrystalChangePositionStatus_ = 0;
+	crystalChange_ = 0;
+	crystalChangeCWLimitStatus_ = 0;
+	crystalChangeCCWLimitStatus_ = 0;
+	regionAStatus_ = 0;
+	regionBStatus_ = 0;
+
+	// Initialize inherited variables.
+
+	setEnumStates(QStringList() << "A" << "B" << "None");
+	setMoveEnumStates(QStringList() << "A" << "B");
+	setAllowsMovesWhileMoving(false);
+	setContextKnownDescription("Region Control");
+
+	// Make connections
+
+	connect( this, SIGNAL(connected(bool)), this, SIGNAL(enumChanged()) );
+	connect( this, SIGNAL(valueChanged(double)), this, SIGNAL(enumChanged()) );
+	connect( this, SIGNAL(setpointChanged(double)), this, SIGNAL(enumChanged()) );
+}
+
+BioXASSSRLMonochromatorRegionControl::~BioXASSSRLMonochromatorRegionControl()
+{
+
+}
+
+double BioXASSSRLMonochromatorRegionControl::minimumValue() const
+{
+	return BioXASSSRLMonochromator::Region::None;
+}
+
+double BioXASSSRLMonochromatorRegionControl::maximumValue() const
+{
+	return BioXASSSRLMonochromator::Region::B;
+}
+
+bool BioXASSSRLMonochromatorRegionControl::canMove() const
+{
+	bool exists = false;
+	bool connected = false;
+	bool canMove = false;
+
+	exists = (
+		upperSlit_ &&
+		lowerSlit_ &&
+		slitsStatus_ &&
+		paddle_ &&
+		paddleStatus_ &&
+		keyStatus_ &&
+		brakeStatus_ &&
+		bragg_ &&
+		braggAtCrystalChangePositionStatus_ &&
+		crystalChange_ &&
+		crystalChangeCWLimitStatus_ &&
+		crystalChangeCCWLimitStatus_ &&
+		regionAStatus_ &&
+		regionBStatus_
+	);
+
+	if (exists) {
+		connected = isConnected();
```

Comment:
It's possible as written that some of the controls won't be set properly; the region control shouldn't move if this is the case. I'm not sure that all of the child controls can actually move (I'm thinking of the status controls in particular), and I didn't want that to impact the overall canMove state of the region control. But there is definitely a problem if moving parts (slit motors, paddle motor, bragg motor, and crystal change motor controls) can't move.